go/vt/mysqlctl/builtinbackupengine.go

```diff
@@ -516,66 +516,37 @@ func (be *BuiltinBackupEngine) ExecuteRestore(
 	restoreConcurrency int,
 	hookExtraEnv map[string]string) (mysql.Position, error) {
 
-	var bh backupstorage.BackupHandle
+	zeroPosition := mysql.Position{}
 	var bm builtinBackupManifest
-	var toRestore int
 
-	for toRestore = len(bhs) - 1; toRestore >= 0; toRestore-- {
-		bh = bhs[toRestore]
-		rc, err := bh.ReadFile(ctx, backupManifest)
-		if err != nil {
-			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
-			continue
-		}
-
-		err = json.NewDecoder(rc).Decode(&bm)
-		rc.Close()
-		if err != nil {
-			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage (cannot JSON decode MANIFEST: %v)", bh.Name(), dir, err)
-			continue
-		}
-
-		logger.Infof("Restore: found backup %v %v to restore with %v files", bh.Directory(), bh.Name(), len(bm.FileEntries))
-		break
-	}
-	if toRestore < 0 {
-		// There is at least one attempted backup, but none could be read.
-		// This implies there is data we ought to have, so it's not safe to start
-		// up empty.
-		return mysql.Position{}, ErrNoCompleteBackup
-	}
-
-	// Starting from here we won't be able to recover if we get stopped by a cancelled
-	// context. Thus we use the background context to get through to the finish.
-
-	logger.Infof("Restore: shutdown mysqld")
-	err := mysqld.Shutdown(context.Background(), cnf, true)
+	bh, err := findBackupToRestore(ctx, cnf, mysqld, logger, dir, bhs, &bm)
 	if err != nil {
-		return mysql.Position{}, err
+		return zeroPosition, err
 	}
 
-	logger.Infof("Restore: deleting existing files")
-	if err := removeExistingFiles(cnf); err != nil {
-		return mysql.Position{}, err
+	// mark restore as in progress
+	if err = createStateFile(cnf); err != nil {
```

Comment:
Now that we remove the state file one level up in `mysqlctl.Restore()`, I was thinking we'd also create it there so all the logic is in one place. That does mean that we'd potentially mark the data dir invalid just because we failed to find a valid backup, but that may be easy to avoid now that you've abstracted `findBackupToRestore()`. We could call that from `mysqlctl.Restore()` instead of here, and then create the state file there as well if we find one, before passing the found backup to `ExecuteRestore()`. What do you think?

Reply:
That would be ideal, but there is a little bit of difficulty with moving the call to `findBackupToRestore` into `mysqlctl.Restore`. The manifest format is different for the two backup methods - builtin and xtrabackup, and we are passing in the appropriate one to `findBackupToRestore` when we call it. Any suggestions on how we could work around that?

Reply:
Ah I didn't notice that there's an implementation-specific arg to `findBackupToRestore`. In that case, it's probably not worth doing surgery on it in this PR.